motionplan/constraint.go

```diff
@@ -10,120 +10,226 @@ import (
+	err := ci.resolveInputsToPositions()
 	if err != nil {
 		return false, nil
 	}
-	steps := PathStepCount(ci.StartPos, ci.EndPos, resolution)
+	steps := PathStepCount(ci.StartPosition, ci.EndPosition, resolution)
 
 	var lastGood []referenceframe.Input
-	// Seed with just the start position to walk the path
-	interpC := &ConstraintInput{Frame: ci.Frame}
-	interpC.StartInput = ci.StartInput
-	interpC.EndInput = ci.StartInput
 
-	for i := 1; i <= steps; i++ {
+	for i := 0; i <= steps; i++ {
 		interp := float64(i) / float64(steps)
-		interpC, err = cachedInterpolateInput(ci, interp, interpC.EndInput, interpC.EndPos)
+		interpConfig := referenceframe.InterpolateInputs(ci.StartConfiguration, ci.EndConfiguration, interp)
+		interpC := &StateInput{Frame: ci.Frame, Configuration: interpConfig}
+		err = interpC.resolveInputsToPositions()
 		if err != nil {
 			return false, nil
 		}
-		pass, _, _ := c.CheckConstraints(interpC)
+		pass, _ := c.CheckStateConstraints(interpC)
 		if !pass {
-			if i > 1 {
-				return false, &ConstraintInput{StartInput: lastGood, EndInput: interpC.StartInput}
+			if i == 0 {
+				// fail on start pos
+				return false, nil
 			}
-			// fail on start pos
-			return false, nil
+			return false, &SegmentInput{StartConfiguration: ci.StartConfiguration, EndConfiguration: lastGood}
 		}
-		lastGood = interpC.StartInput
-	}
-	// extra step to check the end
-	if err != nil {
-		return false, nil
-	}
-	pass, _, _ := c.CheckConstraints(&ConstraintInput{
-		StartPos:   ci.EndPos,
-		EndPos:     ci.EndPos,
-		StartInput: ci.EndInput,
-		EndInput:   ci.EndInput,
-		Frame:      ci.Frame,
-	})
-	if !pass {
-		return false, &ConstraintInput{StartInput: lastGood, EndInput: interpC.StartInput}
+		lastGood = interpC.Configuration
 	}
 
 	return true, nil
 }
 
-// AddConstraint will add or overwrite a constraint function with a given name. A constraint function should return true
-// if the given position satisfies the constraint.
-func (c *constraintHandler) AddConstraint(name string, cons Constraint) {
-	if c.constraints == nil {
-		c.constraints = map[string]Constraint{}
+// CheckSegmentAndStateValidity will check an segment input and confirm that it 1) meets all segment constraints, and 2) meets all
+// state constraints across the segment at some resolution. If it fails an intermediate state, it will return the shortest valid segment,
+// provided that segment also meets segment constraints.
+func (c *ConstraintHandler) CheckSegmentAndStateValidity(cInput *SegmentInput, resolution float64) (bool, *SegmentInput) {
+	valid, subSegment := c.CheckStateConstraintsAcrossSegment(cInput, resolution)
+	if !valid {
+		if subSegment != nil {
+			subSegmentValid, _ := c.CheckSegmentConstraints(subSegment)
+			if subSegmentValid {
+				return false, subSegment
+			}
+		}
+		return false, nil
 	}
-	if cons != nil {
-		c.constraints[name] = cons
+	// all states are valid
+	valid, _ := c.CheckSegmentConstraints(cInput)
```

Comment:
can just return this function call

Reply:
That's not so; the `_` we skip over here is not an err.